### PR TITLE
Update pre-commit hooks and drop support for Python 3.6 & 3.7

### LINKED ISSUE #1108: Pre-commit hooks are not working correctly
Currently, trying to setup the code for development does not work, there are several issues related to outdated pre-commit hooks.

### PR DESCRIPTION
**Pull request by odelalleau**
Fixes #1108

This PR updates pre-commit hooks to the latest current versions of isort / black / flake8 / mypy, and ensures we can run all these tools in the repository without any complaint.

### CODE REVIEW

**Comment by odelalleau** [comment]
@omry any objection to bumping the min version requirement to 3.8? Recent versions of black, etc. have all moved to 3.8.

**Comment by odelalleau** [comment]
Converting to draft until the move to 3.8 is approved

**Comment by odelalleau** [comment]
@omry thoughts on enforcing Python >= 3.8?

**Comment by omry** [comment]
lgtm.
FYI: if you want my attention ping me on a different channel.

File: omegaconf/_utils.py
Comment (omry) [inline omegaconf/_utils.py:643]: This gets into actual cleanup of 3.6 code.
Jasha, did we officially removed support for it yet?

Comment (odelalleau) [reply omegaconf/_utils.py:643]: > This gets into actual cleanup of 3.6 code.

When you said "lgtm" I thought you were saying "yes" to my question "thoughts on enforcing Python >= 3.8?"

> @Jasha10 , did we officially removed support for it yet?

All official releases of OmegaConf / Hydra still support Python 3.6. I'm proposing to drop this support in the next release (because it will make our lives easier, and I doubt anyone still runing 3.6/3.7 would care about not being able to run the latest version).

**Comment by odelalleau** [comment]
Un-drafting as I believe it's ready now.

Next step on my side: doing something similar in Hydra (may take a few more weeks as I'm working on it quite sparingly).

**Comment by odelalleau** [comment]
@Jasha10 & @omry I would like to get at least one official approval on the updated PR before merging -- thanks! (no rush)

**Review by Jasha10 — APPROVED** [review]
Sorry for the slow review. This diff looks good to me!

**Pull request merged**
